src/main/java/uk/ac/ic/wlgitbridge/bridge/Bridge.java

```diff
@@ -326,14 +319,24 @@ public ProjectRepo getUpdatedRepo(
             Optional<Credential> oauth2,
             String projectName
     ) throws IOException, GitUserException {
-        try (LockGuard __ = lock.lockGuard(projectName)) {
+        ProjectContext context = ContextStore.Instance().getContextForProject(projectName);
+        ProjectLock lock = context.getLock();
+        lock.lock();
+        try {
             Optional<GetDocResult> maybeDoc = snapshotAPI.getDoc(oauth2, projectName);
             if (!maybeDoc.isPresent()) {
                 throw new RepositoryNotFoundException(projectName);
             }
             GetDocResult doc = maybeDoc.get();
             Log.info("[{}] Updating repository", projectName);
-            return getUpdatedRepoCritical(oauth2, projectName, doc);
+            ProjectRepo repo = getUpdatedRepoCritical(oauth2, projectName, doc);
+            return repo;
+        } catch (Throwable t) {
+            lock.fail();
+            throw t;
+        } finally {
+            lock.unlock();
+            ContextStore.Instance().clearLocalContexts();
```

Comment:
As mentioned before, I think it's only the one context that we need to clear.

Reply:
Discussed in call:

Actually, this is correct. We sometimes need to lock multiple projects at once, when we are migrating things around.